### DISCUSSION TITLE
allowedDependencies / NamedInterface

### DISCUSSION THREAD

**Discussion by Platzdoch** [spring-projects/spring-modulith#805] [Q&A]
Hello, I find development with Spring Modulith very exciting and am currently rewriting an existing project. This involves resolving many dependencies and sharing services. However, I am having problems allowing dependencies. TheVerify method always throws errors.

I want to allow the service "userService" from the user module in the customer module.

I created a packet info for each one.
directly in the packetService

For user:
@org.springframework.modulith.NamedInterface("userService")
package app.user.rep.service;

For customer
@org.springframework.modulith.ApplicationModule(
        allowedDependencies = "user:: userService"
)

Unfortunately, I keep getting the error message
CustomerResource has parameter of type UserService.
What am I doing wrong? Thanks in advance for your help.

module user:
user 
       - user.a.java
       - user.b.java
   user.rep
         - user.rep.a.java
         - user.rep.b.java
    user.repo.service
                 - user.repo.service.userService
                 package-info.java

module customer:
customer 
   customer.controller
       - customer.resource.java
       package-info.java
   customer.rep
   customer.model

**Comment by Shriharsh-Deshmukh** [comment]
To resolve the issue with Spring Modulith dependency injection:

1. **Correct the `allowedDependencies`:** Ensure the format matches the module structure:
   ```java
   @org.springframework.modulith.ApplicationModule(
   allowedDependencies = "user::userService"
   )
   ```

2. **Check Package Naming:** Verify that the package name in `@NamedInterface("userService")` matches exactly with your `UserService` class location.

3. **Adjust Service Exposure:** Ensure the `userService` is properly exposed via a public interface in the user module.

4. **Validate Dependency Order:** Ensure the `user` module is correctly listed as a dependency in the `customer` module in your `build.gradle` or `pom.xml`.

Recheck these and it should resolve your dependency issue.

**Reply by Platzdoch** [reply]
Thank you for the quick reply. Everything should be fine in the module. See info from the verify method.


# Customer
> Logical name: customer
> Base package: .package.customer
> Direct module dependencies: securityAndUser

# User
> Logical name: user
> Base package: package.user
> Named interfaces:
  + NamedInterface: name=<<UNNAMED>>, types=[]
  + NamedInterface: name=userService, types=[ e.m.c.s.u.r.s.EventService, e.m.c.s.u.r.s.RoleService, e.m.c.s.u.r.s.UserService ]

So far everything should be working correctly. If the service is not recognized, an error message will appear. Unfortunately I have not found any information on how to implement point 4. What needs to be entered?

**Reply by Shriharsh-Deshmukh** [reply]
**What to Enter**
Here’s how you can validate the dependency order in two common build systems:

For Gradle (build.gradle): In the customer module’s build.gradle file, add the following line inside the dependencies block:

**groovy**

```
dependencies {
    implementation project(':user') // Ensure this points to your 'user' module
}
```
This tells Gradle that the customer module depends on the user module.

For Maven (pom.xml): In the customer module’s pom.xml file, add a dependency block like this:

**xml**

```
<dependency>
    <groupId>your.group.id</groupId>  <!-- Replace with your actual group ID -->
    <artifactId>user-module</artifactId>  <!-- Replace with the actual artifact ID of the user module -->
    <version>1.0.0</version>  <!-- Replace with the correct version -->
</dependency>
```
This tells Maven that the  customer module relies on the user module.

**Comment by odrotbohm** [comment]
Any chance you provide a simple reproducer that shows the verification error appearing?
